lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/TaskMessageObserver.cs

```diff
@@ -0,0 +1,134 @@
+
+        public TaskMessageObserver(StreamingNetworkService<GeneralGroupCommunicationMessage> networkService)
+        {
+            _networkService = networkService;
+        }
+
+        /// <summary>
+        /// Registers a node associated with the Task.
+        /// </summary>
+        public void RegisterNodeObserver<T>(NodeMessageObserver<T> observer)
+        {
+            _observers.Add(NodeObserverIdentifier.FromObserver(observer), observer);
+        }
+
+        /// <summary>
+        /// This is called directly from the observer container with the registered IPEndpoint
+        /// of the Task ID.
+        /// </summary>
+        public void OnNext(NsMessage<GeneralGroupCommunicationMessage> value)
+        {
+            Handle(value);
+        }
+
+        /// <summary>
+        /// Handles the group communication message.
+        /// </summary>
+        private void Handle(NsMessage<GeneralGroupCommunicationMessage> value, bool isRegistration = false)
+        {
+            // This is mainly used to handle the case should ObserverContainer
+            // decide to trigger handlers concurrently for a single message.
+            if (isRegistration)
+            {
+                // Process the registration message
+                _registrationMessage = value;
+            }
+            else if (_registrationMessage != null && value == _registrationMessage)
+            {
+                // This means that we've already processed the message.
+                // Ignore this message and discard the reference.
+                _registrationMessage = null;
+                return;
+            }
+
+            var gcMessage = value.Data.First();
+
+            IObserver<NsMessage<GeneralGroupCommunicationMessage>> observer;
+            if (!_observers.TryGetValue(NodeObserverIdentifier.FromMessage(gcMessage), out observer))
+            {
+                throw new InvalidOperationException();
+            }
+
+            observer.OnNext(value);
+        }
+
+        /// <summary>
+        /// This is called from the universal observer in ObserverContainer for the first message.
+        /// </summary>
+        public void OnNext(IRemoteMessage<NsMessage<GeneralGroupCommunicationMessage>> value)
+        {
+            // Lock to prevent duplication of messages.
+            lock (_registrationLock)
+            {
+                if (_hasRegistered)
+                {
+                    return;
+                }
+
+                var socketRemoteId = value.Identifier as SocketRemoteIdentifier;
+                if (socketRemoteId == null)
+                {
+                    throw new InvalidOperationException();
+                }
+
+                // Handle the message first, then register the observer.
+                Handle(value.Message, true);
+                _networkService.RemoteManager.RegisterObserver(socketRemoteId.Addr, this);
```

Comment:
It will only be called from universal observer for the first message, which is then de-duped because although right now there are no plans for it, it's possible that observers can be called in parallel, and the bootstrapped observer can thus be in a "race" with the universal observer. This can be solved if we can identify the endpoints before the first message, but sadly we are not able to do it yet.